ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala

```diff
@@ -0,0 +1,187 @@
+    val party = partyRecordUpdate.party
+    val attemptedUpdateF = withState(
+      state.get(party) match {
+        case Some(info) =>
+          val updatedInfo: PartyRecordInfo = doUpdatePartyRecord(
+            partyRecordUpdate = partyRecordUpdate,
+            party = party,
+            info = info,
+          )
+          Right(toPartyRecord(updatedInfo))
+        case None =>
+          throw PartyRecordNotFoundOnUpdateException
+      }
+    ).map(tapSuccess { updatePartyRecord =>
+      logger.info(s"Updated party record in a participant local store: ${updatePartyRecord}")
+    })
+    attemptedUpdateF.recoverWith[Result[domain.ParticipantParty.PartyRecord]] {
+      case PartyRecordNotFoundOnUpdateException =>
+        for {
+          partyExistsOnLedger <- ledgerPartyExists.exists(party)
+          createdPartyRecord <-
+            if (partyExistsOnLedger) {
+              withState {
+                val newPartyRecord = domain.ParticipantParty.PartyRecord(
+                  party = party,
+                  metadata = domain.ObjectMeta.empty,
+                )
+                for {
+                  _ <- withoutPartyRecord(party = newPartyRecord.party) {
+                    doCreatePartyRecord(newPartyRecord)
+                  }
+                  _ <- withPartyRecord(party) { info =>
+                    val updatedInfo: PartyRecordInfo = doUpdatePartyRecord(
+                      partyRecordUpdate = partyRecordUpdate,
+                      party = party,
+                      info = info,
+                    )
+                    toPartyRecord(updatedInfo)
+                  }
+                  updatePartyRecord <- {
+                    withPartyRecord(party) { updatedInfo =>
+                      toPartyRecord(updatedInfo)
+                    }
+                  }
+                } yield updatePartyRecord
+              }.map(tapSuccess { newPartyRecord =>
+                logger.error(
+                  s"Created a new party record in a participant local store: ${newPartyRecord}"
+                )
+              })(scala.concurrent.ExecutionContext.parasitic)
+            } else {
+              Future.successful(
+                Left(ParticipantPartyRecordStore.PartyNotFound(party))
+              )
+            }
+        } yield createdPartyRecord
+    }
+  }
+
+  private def doUpdatePartyRecord(
+      partyRecordUpdate: PartyRecordUpdate,
+      party: Party,
+      info: PartyRecordInfo,
+  ): PartyRecordInfo = {
+    val existingAnnotations = info.annotations
+    val updatedAnnotations =
+      partyRecordUpdate.metadataUpdate.annotationsUpdateO.fold(existingAnnotations) {
+        case AnnotationsUpdate.Merge(newAnnotations) => existingAnnotations.concat(newAnnotations)
+        case AnnotationsUpdate.Replace(newAnnotations) => newAnnotations
+      }
+    val updatedInfo = PartyRecordInfo(
+      party = party,
+      resourceVersion = info.resourceVersion + 1,
```

Comment:
There is no gurarantee partyRecordUpdate has the correct resourceVersion

Reply:
Fixing